api/server/router/container/backend.go

```diff
@@ -43,7 +43,7 @@ type stateBackend interface {
 	ContainerStop(name string, seconds int) error
 	ContainerUnpause(name string) error
 	ContainerUpdate(name string, hostConfig *container.HostConfig) ([]string, error)
-	ContainerWait(name string, timeout time.Duration) (int, error)
+	ContainerWait(name string, timeout time.Duration, waitForRemoval bool) (int, error)
```

Comment:
So what's the conclusion now? I'm a little bit confused on the next step. 

I feels that I'm not making myself very clear, so I'll say more word about this:
If container is started with a TTY, then it can be detached with detach keys. In such case, if we choose not to wait it to stop but get the exit code directly, we may get wrong exit code due to its slow shutdown;  If we choose to wait it to stop, it may be detached but not stopped so the wait hang, what's worse is that if the container is restarted based on restart policy too quickly, the ContainerWait can also hang because it's restored to running state from stopped state.  So we need to add a timeout if we choose to wait the container in case to hang. The drawbacks is, detach from container will still wait 2 seconds before client exits.

There's no difference if we wait `die` (plus `destroyed`) event from new `wait` API. 

The benefit of using event is that we don't need to modify the `wait` API, but actually I think it will  need more codes and won't be a more elegant implementation. 

That's why I think the modified `wait` API is better, and I believe we can't find a single implementation which satisfy everyone.  But if @cpuguy83 is really against this, I can change the implement, it's also workable I think. 

So what's your idea? @mlaventure @cpuguy83